### PR TITLE
Convert charge deposition

### PR DESCRIPTION
**Pull request by dpgrote**
This implements the charge deposition in c++, including RZ single mode.

Note that this has not yet been tested on GPU

### CODE REVIEW

File: Source/Particles/WarpXParticleContainer.cpp [hunk 1]
```diff
@@ -552,140 +553,86 @@ WarpXParticleContainer::DepositCurrent(WarpXParIter& pti,
...
+        tilebox = amrex::coarsen(pti.tilebox(),ref_ratio);
+    }
+    
+    tilebox.grow(ngRho);
 
 #ifdef AMREX_USE_GPU
-      data_ptr = (*rhomf)[pti].dataPtr(icomp);
-      auto rholen = (*rhomf)[pti].length();
+    // No tiling on GPU: rho_arr points to the full rho array.
+    Array4<Real> const& rho_arr = rho->array(pti);
```
Comment (RemiLehe) [inline Source/Particles/WarpXParticleContainer.cpp]: It looks like the above line should depend on `icomp`, correct?
Could you test the code on GPU, just to make sure that, with the above fix, the code gives correct results?

Comment (dpgrote) [reply Source/Particles/WarpXParticleContainer.cpp]: You're correct. Do you know how to do that? Would array take a second argument?
I'm working on getting access to corigpu - I only had it for the hackathon.

Comment (MaxThevenet) [reply Source/Particles/WarpXParticleContainer.cpp]: Some possible solution is given in the AMReX doc:
https://amrex-codes.github.io/amrex/docs_html/GPU.html#launching-c-nested-loops

Comment (dpgrote) [reply Source/Particles/WarpXParticleContainer.cpp]: I pushed a change that should fix the issue.

**Comment by RemiLehe** [comment]
Thanks! I just pushed a small syntax fix, and tested the code on GPU (in the case `Langmuir_multi_psatd`, which uses the charge deposition as a source for the Maxwell pusher), and the test passes.

**Comment by dpgrote** [comment]
Thanks!

**Pull request merged**